File: src/com/dafttech/terra/game/world/items/persistence/Prototype.java
package com.dafttech.terra.game.world.items.persistence;

import java.lang.reflect.Field;
import java.util.HashMap;

public class Prototype {
    public String className = "";
    public HashMap<Field, Object> values = new HashMap<Field, Object>();

    @Override
    public int hashCode() {
        return getHashBase().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Prototype) return hashCode() == ((Prototype) obj).hashCode();
        if (obj instanceof GameObject) return hashCode() == ((GameObject) obj).hashCode();

        return true;
    };

    public String getHashBase() {
        StringBuilder hashBuilder = new StringBuilder();

        hashBuilder.append(className);

        for (Object f : values.values()) {
            hashBuilder.append(f.toString());
        }

        return hashBuilder.toString();
    }

    public GameObject toGameObject() {
        GameObject object = null;

        try {
            @SuppressWarnings("unchecked")
            Class<GameObject> cl = (Class<GameObject>) Class.forName(className);

            object = cl.newInstance();

            Field[] fields = cl.getFields();

            for (Field f : fields) {
                if (f.isAnnotationPresent(Persistent.class)) f.set(object, f);
            }
        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }

        return object;
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/items/persistence/Prototype.scala
package com.dafttech.terra.game.world.items.persistence

import java.lang.reflect.Field

import com.dafttech.terra.game.world.items.TileItem
import com.dafttech.terra.game.world.tiles.Tile

class Prototype {
  var className = ""
  private var values = Map[Field, AnyRef]()

  override def hashCode: Int = getHashBase.hashCode

  override def equals(obj: Any): Boolean = obj match {
    case _: Prototype => hashCode == obj.hashCode
    case _: GameObject => hashCode == obj.hashCode
    case _ => true
  }

  def getHashBase: String = className + values.values.map(_.toString).mkString

  def addValue(field: Field, value: AnyRef): Unit = values = values + (field -> value)

  def toGameObject: GameObject = {
    Class.forName(className) match {
      case tileItem if tileItem == classOf[TileItem] =>
        TileItem(values(TileItem.tileField).asInstanceOf[Tile])

      case clazz =>
        @SuppressWarnings(Array("unchecked"))
        val cl = clazz.asInstanceOf[Class[GameObject]]
        val obj = cl.newInstance()
        for ((field, value) <- values) field.set(obj, value)
        obj
    }
  }
}